lib/Agrammon/OutputFormatter/CSV.pm6

```diff
@@ -1,46 +1,97 @@
                 take @data.join(';');
+                if $include-filters {
+                    if $raw-value ~~ Agrammon::Outputs::FilterGroupCollection && $raw-value.has-filters {
+                        add-filters($raw-value, $module, $var, $unit, $prefix, :$all-filters);
+                    }
+                }
             }
-         }
-        else {
-            for .value.sort(*.key) {
-                my $instance-id = .key;
-                for .value.sort(*.key) {
-                    my $fq-name = .key;
+        }
+        when Array {
+            for sorted-kv($_) -> $instance-id, %instance-outputs {
+                my $filters = '';
+                for sorted-kv(%instance-outputs) -> $fq-name, %values {
                     my $q-name = $module ~ '[' ~ $instance-id ~ ']' ~ $fq-name.substr($module.chars);
-                    for .value.sort(*.key) {
+                    for sorted-kv(%values) -> $output, $value {
+                        my $var-print = $model.output-print($fq-name, $output) ~ ',All';
+                        next unless $var-print.split(',') ∩ @print-set;
+
+                        my $raw-value = $value;
+                        my $unit = $model.output-unit($module, $output, $language);
                         my @data = (
-                            $q-name, .key, flat-value(.value) // '',
-                            $model.output-unit($fq-name, .key, $unit-language)
+                            $q-name,
+                            $output,
+                            $filters,
+                            flat-value($raw-value) // '',
+                            $unit
                         );
                         @data.unshift($prefix) if $prefix;
                         take @data.join(';');
+                        if $include-filters {
+                            if $raw-value ~~ Agrammon::Outputs::FilterGroupCollection && $raw-value.has-filters {
+                                add-filters($raw-value, $q-name, $fq-name, $unit, $prefix, :$all-filters);
+                            }
+                        }
                     }
                 }
             }
         }
-    }).join("\n") ~ "\n";
+    }).join("\n");
+}
+
+sub add-filters(Agrammon::Outputs::FilterGroupCollection $collection,
+        $q-name, $fq-name, $unit, Str $prefix, Bool :$all-filters) {
+    my @results = $collection.results-by-filter-group(:all($all-filters));
+    for @results {
+        my %filters := .key;
+        my $raw-value := .value;
+        my $filters = %filters.values.join(',');
+        my @filters = %filters.map: { .key ~ '=' ~ .value };
+        for (@filters || '(Uncategorized)').kv -> $idx, $filter-id {
+            my @data = (
+                $q-name,
+                $fq-name,
+                $filters,
+                flat-value($raw-value) // '',
+                $unit
+            );
+            @data.unshift($prefix) if $prefix;
```

Comment:
Alternatively you can slip it into the original array construction if `|($prefix if $prefix)`

Reply:
Done